### PR TITLE
Remove singleton behavior from numbering

### LINKED ISSUE #43: Question/Idea: Implications of Numbering being a singleton in multi-user rails app
Could there be any issues with the above class being a singleton if say two users generated a docx file at the same time? It seems like the definitions being registers would get fouled up since if I am thinking correctly the rails app would only ever have one instance of Numbering available to work with.

```ruby
    def reset!
      @numid = 1000
      @definitions = []
    end

    def register(style)
      @numid += 1
      definition = Definition.new(@numid, style)
      @definitions << definition
      definition
    end
```

Could it potentially be more robust to register a numbering instance on the Context object being passed around and use that instead? You should be able to maintain the same expected behavior if Context.trasnform is moved out of Document.process and into the template.rb file (likely in the render method).

- [x] Convert Context from module into class
- [ ] Register a local Processor.Numbering instance to the Context instance

### PR DESCRIPTION
**Pull request by stadelmanma**
Fixes #43 

The rebase didn't go smoothly and it was going to cause a big mess so I opted to purge the old branch and do a squashed merge instead. The diff appears rather larger but almost all of the changes are from passing the `env` variable into multiple places where Numbering was previously used, especially in the several HTML tests.

### CODE REVIEW

**Comment by senny** [comment]
@stadelmanma patch conflicts with the current master.

**Comment by stadelmanma** [comment]
@senny sorry about that, forgot I changed some of the inner workings in converter.rb & converter_test.rb. Conflicts have been resolved.

**Comment by senny** [comment]
@stadelmanma would you mind sending me an email? I wanted to contact you but wasn't able to find an email address. Sorry for misusing the PR like this 😅

File: lib/sablon/content.rb [hunk 1]
```diff
@@ -51,7 +52,7 @@ def initialize(value)
         super value.to_s
       end
 
-      def append_to(paragraph, display_node)
+      def append_to(paragraph, display_node, _env)
```
Comment (senny) [inline lib/sablon/content.rb]: I don't think this codebase uses the `_` prefix to mark unused variables. I'm personally not too big of a fan of it. Can we remove it?

File: lib/sablon/content.rb [hunk 2]
```diff
@@ -64,32 +65,44 @@ def append_to(paragraph, display_node)
...
+      def self.id
+        :word_ml
+      end
+
+      def self.wraps?(_)
+        false
+      end
 
-      def append_to(paragraph, display_node)
+      def append_to(paragraph, _display_node, _env)
```
Comment (senny) [inline lib/sablon/content.rb]: Let's remove the `_` prefixes.

File: lib/sablon/environment.rb [hunk 3]
```diff
@@ -3,18 +3,20 @@ module Sablon
...
     def alter_context(context = {})
       new_context = @context.merge(context)
-      Environment.new(@template, new_context)
+      Environment.new(@template, new_context, @numbering)
     end
 
     private
 
-    def initialize(template, context = {})
+    def initialize(template, context = {}, numbering=nil)
```
Comment (senny) [inline lib/sablon/environment.rb]: I think we can assign the default in the signature:

```ruby
def initialize(template, context = {}, numbering=Numbering.new)
```

**Comment by stadelmanma** [comment]
@senny I have made the requested changes. I actually ended up doing a slight refactor on the Environment class so instead of just passing in the numbering reference as needed to `initialize` you can pass in an entire environment. The method then pulls the template and numbering references from that parent environment passed in. 

Looking forward to #54, we will have more things registered to the environment that will need passed on when calling `alter_context`.

File: lib/sablon/content.rb [hunk 4]
```diff
@@ -64,32 +65,44 @@ def append_to(paragraph, display_node)
...
     end
 
+    # handles direct addition of WordML to the document template
     class WordML < Struct.new(:xml)
       include Sablon::Content
-      def self.id; :word_ml end
-      def self.wraps?(value) false end
+      def self.id
+        :word_ml
+      end
```
Comment (senny) [inline lib/sablon/content.rb]: One more thing, any particular reason you decided to reformat these inline methods? I think it does introduce more noise to this diff than necessary. Also, I like them for hardcoded definitions like the ones in these classes.

**Comment by stadelmanma** [comment]
@senny Cleared off the reformats from the diff. Rubocop in my IDE has default settings and I have another project where I strictly follow the linter. I probably absent-mindedly reformatted them at some point in time.

**Comment by senny** [comment]
@stadelmanma thank you 💛

**Pull request merged**